File: word-break.py
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        table = [False] * (len(s) + 1)
        table[len(s)] = True

        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if i + len(w) <= len(s) and s[i : i + len(w)] == w:
                    table[i] = table[i + len(w)]
                if table[i]:
                    break

        return table[0]

File: word-break.cpp
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  bool wordBreak(string s, vector<string> &wordDict) {
    vector<bool> table(s.size() + 1, false);
    table[s.size()] = true;

    for (int i = s.size() - 1; i >= 0; --i) {
      for (string w : wordDict) {
        string ss = s.substr(i, w.size());

        if ((i + w.size() <= s.size()) &&
            count(wordDict.begin(), wordDict.end(), ss)) {
          table[i] = table[i + w.size()];
        }

        if (table[i])
          break;
      }
    }

    return table[0];
  }
};
